Fix Start-Process -Wait with -Credential

jborean93 opened the pull request:
# PR Summary

Allows a non-administrator user to be able to -Wait on a Start-Process call with a custom credential specified.

~~Currently thinking off a way I can test out this scenario.~~ After creating a test to run in CI I've found that it's too hard to create a reliable test that wouldn't block the whole runner on a failing scenario. The test itself also required admin access and was quite complicated to set up all the synchronisation work so I've decided not to include it.

Fixes: https://github.com/PowerShell/PowerShell/issues/17033
Alternative to https://github.com/PowerShell/PowerShell/pull/19082

## PR Context

https://github.com/PowerShell/PowerShell/issues/17033

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/main/reference/7.3/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): <!-- Experimental feature name(s) here -->
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: <!-- Number/link of that issue here -->
- **Testing - New and feature**
    - [x] N/A or can only be tested interactively
    - **OR**
    - [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [x] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and opened an issue in the relevant tool repository. This may include:
        - [ ] Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode
        (which runs in a different PS Host).
            - [ ] Issue filed: <!-- Number/link of that issue here -->
        - [ ] Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
            - [ ] Issue filed: <!-- Number/link of that issue here -->
        - [ ] Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
            - [ ] Issue filed: <!-- Number/link of that issue here -->
        - [ ] Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).
            - [ ] Issue filed: <!-- Number/link of that issue here -->

iSazonov commented on src/Microsoft.PowerShell.Commands.Management/commands/management/Process.cs:
```diff
@@ -744,7 +744,8 @@ private static string RetrieveProcessUserName(Process process)
                 int error;
-                if (!Win32Native.OpenProcessToken(process.Handle, TOKEN_QUERY, out processTokenHandler)) { return null; }
+                if (!Win32Native.OpenProcessToken(process.Handle, TOKEN_QUERY, out processTokenHandler))
+                { return null; }
```
Too many unrelated changes in the PR. Please revert or move to another PR.

jborean93 replied:
Yea sorry that’s vscode autoformating files on save. I’ll revert them before making the PR ready.

jborean93 replied:
These have been reverted with the latest commit.

iSazonov commented on src/Microsoft.PowerShell.Commands.Management/commands/management/Process.cs:
```diff
@@ -2006,15 +2025,65 @@ protected override void BeginProcessing()
+                if (Wait)
+                {
+                    jobObject = new();
+                    jobAssigned = jobObject.AssignProcessToJobObject(processInfo.Process.DangerousGetHandle());
```
Can we use `process.SafeHandle.DangerousGetHandle`? If so we could do minimal changes.

jborean93 replied:
I can most likely have the AssingProcessToJobObject accept a SafeHandle and avoid the DangerousGetHand call. Ultimately it does lead to the same thing though.

iSazonov commented on src/Microsoft.PowerShell.Commands.Management/commands/management/Process.cs:
```diff
@@ -2014,7 +2014,55 @@ protected override void BeginProcessing()
+            }
+            else
+            {
+#if UNIX
+                process = new Process() { StartInfo = startInfo };
+                SetupInputOutputRedirection(process);
+                process.Start();
```
Do you know anything stopping us from direct using process.Start() on Windows too?

I see only limitation - WindowStyle. .Net API support this only for shellexecution. We could ask .Net team to enhance the API and then we could remove the pinvoke at all.

jborean93 replied:
It should be technically possible. I originally assumed the biggest blocker was that the dotnet API didn't support specifying the redirection file paths and the code would have to manually do all that. But it seems like it already does that for the non-Windows side so we could just reuse that. There could be other things that might crop up but I preferred to just address the original problem in question and potentially look at the larger change in the future.

iSazonov replied:
Thanks for confirmation! I will open new issue in .Net repo. If they confirm we can continue in more simple way.

iSazonov replied:
Isuue https://github.com/dotnet/runtime/issues/81681

jborean93 committed:
Fix Start-Process -Wait with -Credential

Allows a non-administrator user to be able to -Wait on a Start-Process
call with a custom credential specified.

SteveL-MSFT commented on src/Microsoft.PowerShell.Commands.Management/commands/management/Process.cs:
```diff
@@ -2014,7 +2014,55 @@ protected override void BeginProcessing()
+                // the ALL_ACCESS token from CreateProcess here to setup the
+                // job object assignment early if -Wait was specified.
+                // https://github.com/PowerShell/PowerShell/issues/17033
+                if (Wait)
```
It's not clear to me why you need to assign the job object here rather than just do it once at line 2098?

SteveL-MSFT reviewed (approved):
Thanks for submitting this!  LGTM.  The changes here helped me understand what exactly you were asking for previously.

jborean93 commented:
@SteveL-MSFT is this waiting on anything?

PaulHigin commented:
@jborean93   Thanks for your contribution!
